src/Symfony/Component/Scheduler/Generator/MessageGenerator.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Scheduler\Generator;
+
+use Psr\Clock\ClockInterface;
+use Symfony\Component\Clock\Clock;
+use Symfony\Component\Scheduler\Schedule;
+use Symfony\Component\Scheduler\ScheduleProviderInterface;
+use Symfony\Component\Scheduler\Trigger\TriggerInterface;
+
+/**
+ * @experimental
+ */
+final class MessageGenerator implements MessageGeneratorInterface
+{
+    private TriggerHeap $triggerHeap;
+    private ?\DateTimeImmutable $waitUntil;
+    private CheckpointInterface $checkpoint;
+
+    public function __construct(
+        private Schedule $schedule,
+        string|CheckpointInterface $checkpoint,
+        private ClockInterface $clock = new Clock(),
+    ) {
+        $this->waitUntil = new \DateTimeImmutable('@0');
+        if (\is_string($checkpoint)) {
+            $checkpoint = new Checkpoint('scheduler_checkpoint_'.$checkpoint, $this->schedule->getLock(), $this->schedule->getState());
+        }
+        $this->checkpoint = $checkpoint;
+    }
+
+    public function getMessages(): \Generator
+    {
+        if (!$this->waitUntil
+            || $this->waitUntil > ($now = $this->clock->now())
+            || !$this->checkpoint->acquire($now)
+        ) {
+            return;
+        }
+
+        $lastTime = $this->checkpoint->time();
+        $lastIndex = $this->checkpoint->index();
+        $heap = $this->heap($lastTime);
+
+        while (!$heap->isEmpty() && $heap->top()[0] <= $now) {
+            /** @var TriggerInterface $trigger */
+            [$time, $index, $trigger, $message] = $heap->extract();
+            $yield = true;
+
+            if ($time < $lastTime) {
+                $time = $lastTime;
+                $yield = false;
+            } elseif ($time == $lastTime && $index <= $lastIndex) {
+                $yield = false;
+            }
+
+            if ($nextTime = $trigger->getNextRunDate($time)) {
+                $heap->insert([$nextTime, $index, $trigger, $message]);
+            }
+
+            if ($yield) {
+                yield $message;
+                $this->checkpoint->save($time, $index);
```

Comment:
I'm wondering here: should we save before or after the yield? Saving after means that if the generator is never resumed (as could happen when an exception occurs), we're going to keep trying the same message. Legit?

Reply:
Yes, it should be retried. The Messnger deals with exceptions from here.